activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,155 @@
+      @model, @connection, @inserts, @on_duplicate, @returning, @unique_by = model, model.connection, inserts, on_duplicate, returning, unique_by
+      @returning = connection.supports_insert_returning? && primary_keys if @returning.nil?
+      @returning = false if @returning == []
+      @on_duplicate = :skip if @on_duplicate == :update && updatable_columns.empty?
+
+      ensure_valid_options_for_connection!
+    end
+
+    def execute
+      if inserts.present?
+        connection.exec_query to_sql, "Bulk Insert"
+      else
+        ActiveRecord::Result.new([], [])
+      end
+    end
+
+    def updatable_columns
+      keys - readonly_columns - unique_by_columns
+    end
+
+    private
+      def ensure_valid_options_for_connection!
+        if returning && !connection.supports_insert_returning?
+          raise ArgumentError, "#{connection.class} does not support :returning"
+        end
+
+        unless %i{ raise skip update }.member?(on_duplicate)
+          raise NotImplementedError, "#{on_duplicate.inspect} is an unknown value for :on_duplicate. Valid values are :raise, :skip, and :update"
+        end
+
+        if on_duplicate == :skip && !connection.supports_insert_on_duplicate_skip?
+          raise ArgumentError, "#{connection.class} does not support skipping duplicates"
+        end
+
+        if on_duplicate == :update && !connection.supports_insert_on_duplicate_update?
+          raise ArgumentError, "#{connection.class} does not support upsert"
+        end
+
+        if unique_by && !connection.supports_insert_conflict_target?
+          raise ArgumentError, "#{connection.class} does not support :unique_by"
+        end
+      end
+
+      def to_sql
+        connection.build_insert_sql(ActiveRecord::InsertAll::Builder.new(self))
+      end
+
+      def keys
+        inserts.present? ? inserts.first.keys.map(&:to_s) : []
+      end
+
+      def readonly_columns
+        primary_keys + model.readonly_attributes.to_a
+      end
+
+      def unique_by_columns
+        unique_by ? unique_by.fetch(:columns).map(&:to_s) : []
+      end
+
+      def primary_keys
+        Array.wrap(model.primary_key)
+      end
+
+
+      class Builder
+        attr_reader :model
+
+        def initialize(insert_all)
+          @insert_all, @model, @connection = insert_all, insert_all.model, insert_all.connection
+        end
+
+        def skip_duplicates?
+          insert_all.on_duplicate == :skip
+        end
+
+        def update_duplicates?
+          insert_all.on_duplicate == :update
+        end
+
+        def into
+          [ "INTO ", model.quoted_table_name, "(", columns_list, ") " ].join
+        end
+
+        def values_list
+          columns = connection.schema_cache.columns_hash(model.table_name)
+
+          values_list = insert_all.inserts.map do |attributes|
+            attributes = attributes.stringify_keys
+
+            unless attributes.keys.to_set == keys.to_set
```

Comment:
Calling `keys` on self twice in this loop seems needlessly wasteful. Especially since it creates an array and then maps it to strings.

We could do:

```ruby
columns = …
keys    = insert_all.keys

values_list = …
```

Or have insert_all memoize its keys.

Reply:
We should also add a validation in InsertAll that the "master" keys are uniq, so we don't have to use `keys.to_set` here.

We might even want to push this entire validation into InsertAll ala:

```ruby
@keys = validate_keys(inserts)

def validate_keys(inserts)
  inserts.first.keys.map(&:to_s).uniq.tap do |keys|
    if inserts.any? { |insert| (insert.keys - keys).any? } # Aims to abort early. Catches non-unique keys in the first insert too.
      raise ArgumentError, "All inserts must have the same keys" # Simplified exception message too.
    end
  end
end
```